Mathlib/Topology/Instances/Rat.lean

```diff
@@ -118,3 +118,21 @@ nonrec theorem totallyBounded_Icc (a b : ℚ) : TotallyBounded (Icc a b) := by
       (totallyBounded_Icc (a : ℝ) b)
 
 end Rat
+
+namespace NNRat
+
+instance : MetricSpace ℚ≥0 :=
+  Subtype.metricSpace
+
```

Comment:
```suggestion

theorem dist_eq (p q : ℚ≥0) : dist p q = dist (p : ℚ) (q : ℚ) := rfl
theorem nndist_eq (p q : ℚ≥0) : dist p q = nndist (p : ℚ) (q : ℚ) := rfl

```

Reply:
Could even add
```lean
theorem sphere_eq (q : ℚ≥0) (r : ℝ) : Metric.sphere q r = (↑) ⁻¹' Metric.sphere (q : ℚ) r := rfl
theorem ball_eq (q : ℚ≥0) (r : ℝ) : Metric.ball q r = (↑) ⁻¹' Metric.ball (q : ℚ) r := rfl
theorem closedBall_eq (q : ℚ≥0) (r : ℝ) : Metric.closedBall q r = (↑) ⁻¹' Metric.closedBall (q : ℚ) r := rfl
```
but perhaps that's overkill

Reply:
We have `Subtype.dist_eq` and `Subtype.nndist_eq` as rfl lemmas. Are those enough or do you want the specialized versions? Should those be `@[simp]` tagged so that they're dsimp-discoverable?

Would you like me to add the `sphere/ball/closedBall` lemmas for `Subtype` or specifically NNRat?

Reply:
> ```suggestion
> 
> theorem dist_eq (p q : ℚ≥0) : dist p q = dist (p : ℚ) (q : ℚ) := rfl
> theorem nndist_eq (p q : ℚ≥0) : nndist p q = nndist (p : ℚ) (q : ℚ) := rfl
> 
> ```

I would make those simp and norm_cast backwards

Reply:
Those aren't enough, because IIRC NNRat is not reducible, and so those lemmas won't apply.

It's fine to leave the sphere etc lemmas to a future PR, especially if you add them for Subtype + NNReal as well

Reply:
I left them stated in the `dist_eq` direction but with
```
@[simp ←, push_cast]
```
Is that appropriate?

Reply:
Does the `push_cast` attribute even exist?

Reply:
Yes:
```
import Mathlib

instance : MetricSpace ℚ≥0 :=
  Subtype.metricSpace

@[simp ←, push_cast]
lemma dist_eq (p q : ℚ≥0) : dist p q = dist (p : ℚ) (q : ℚ) := rfl
@[simp ←, push_cast]
lemma nndist_eq (p q : ℚ≥0) : nndist p q = nndist (p : ℚ) (q : ℚ) := rfl

example (x y : ℚ≥0) : dist (x : ℚ) y = dist x y := by
  norm_cast

example (x y : ℚ≥0) : dist (x : ℚ) y = dist x y := by
  simp? says simp only [← dist_eq]
```

Reply:
🤯

Reply:
Isn't `push_cast` the same as `norm_cast`? That is, the latter auto-detects that `push_cast` is appropriate?

Reply:
No, the attributes don't work the same way:
```
import Mathlib

instance : MetricSpace ℚ≥0 :=
  Subtype.metricSpace

@[push_cast]
lemma dist_eq (p q : ℚ≥0) : dist p q = dist (p : ℚ) (q : ℚ) := rfl
@[norm_cast] -- norm_cast: badly shaped lemma, lhs must contain at least one coe dist p q
lemma dist_eq' (p q : ℚ≥0) : dist p q = dist (p : ℚ) (q : ℚ) := rfl
```

Reply:
> Isn't `push_cast` the same as `norm_cast`? That is, the latter auto-detects that `push_cast` is appropriate?

This sounds like lost Lean 3 behaviour